### ISSUE TITLE
[amp-story-player] Fetch more stories from endpoint

### ISSUE THREAD

**Issue by Enriqe** [ampproject/amphtml#29864]
We should allow publishers to specify an endpoint where more stories could be loaded once the user reaches the end of stories currently available in the client side.

The publisher configuration will look like the following:

```
<amp-story-player>
  <script type="application/json">
    {
      "behavior": {
        "on": "end", // for future-proofing
        "action": "fetch", // or circular-wrapping
        "distance": "2", // could come later as a new feature (controls how far in advance the stories should be loaded)
        ...
      },
       ...
    }
  </script>
</amp-story-player>
```

The endpoint should return a JSON with the following structure to represent a story:

```
{
  {
    href: string,
    title: ?string,
    poster: ?string, 
  },
 ...
}

```

@ampproject/wg-stories

**Comment by gmajoulet** [comment]
Can you expand the example to include pagination params? At a minimum we should have an `offset` parameter.
I think we should encapsulate the stories into a `{data: {stories: []}` or anything maybe, to make things more future proof.

**Comment by newmuis** [comment]
Honestly, the request should probably be parameterized like `amp-analytics`; maybe we can reuse their variable substitution.  But it's quite possible different backenda would want different signals to know what to provide next.

**Comment by Enriqe** [comment]
What do you think of something like:

```
<amp-story-player on-end="fetch" endpoint="www.example.com/endpoint.json?offset=${offset}&length=${numStories}">
    ...
</amp-story-player>
```

This would retrieve a subset of stories of size `length` and starting with the `offset` value. The player would then increment the `offset` by the given `numStories` value on each outgoing request.

The returning payload would look like the following:

```
"data": {
  "stories": [
    {
      "href": string,
      "title": ?string,
      "poster": ?string, 
    },
    ...
  ]
}
```

#### Example 1

```
<amp-story-player on-end="fetch" endpoint="www.example.com/endpoint.json?offset=0&length=3">
    ...
</amp-story-player>
```

When the user finishes consuming the existing stories in the player (or whatever heuristic we choose, could be when in story N-1), it would request 3 more starting at offset 0. After that, it will request 3 more starting from offset 3.

**Comment by Enriqe** [comment]
We should also remember to think of how this would work for the AMP version. I'll look into @newmuis' suggestion to see if we can reuse `amp-analytics`' variable substitution for that.

**Comment by Enriqe** [comment]
Expanding some more options:

### Option A

* Publisher provides an endpoint, offset, and length attributes on the player element. 
* We define what the query parameters should be called and document it.

```
<amp-story-player endpoint="https://www.example.com/endpoint.php" offset="0" length="3">
</amp-story-player>
```

### Option B (Preferred)

* Publisher provides an endpoint with their own defined attributes in a templated way for us to make the variable substitution.
* Publisher defines query parameters names.

```
<amp-story-player endpoint="https://www.example.com/endpoint.php?pageNumber=${offset}&numStories=${length}" offset="0" length="3">
</amp-story-player>
```

### Option C

* More advanced/complex configuration options. 
* Publisher defined query parameters.
* We could fit other configuration options into this object (circular wrapping, UI customization, etc...)

```
<amp-story-player>
  <script type="application/json">
    {
      "onEndOfStories": {
        "behavior": "fetch",
        "endpoint": "https://example.com/endpoint.php&page=${offset}&numStories=${length}",
        "vars": {
            "offset": "0",
            "length": "3"
        }
      }
    }
  </script>
</amp-story-player>
```

IMHO I think we could keep it simple and just require an endpoint and two additional attributes for the pagination. But if we wanted to we could provide more advanced/feature-proof features that could be combined with a more general configuration object (could be merged into the UI customization JSON, or other end of stories behaviors / events). WDYT  @gmajoulet @newmuis ?

**Comment by newmuis** [comment]
Nice thing about B is that it becomes easier for publishers to personalize the response on things that we haven't thought about.  For example, if publishers want to serve a different set of stories based on users' engagement.

With option A, they'd need to file a FR and we'd need to implement each experience E2E.  With option B, worst case, we'd need to only implement each new signal that they need (best case, they can combine existing signals and need no work from us).

I think option C is a bit of a different question: should we have an HTML-based config or a JSON-based config?  We could do either one with option A or option B.  I think my (weak?) preference is JSON-based config with the URL variable substitution (which is roughly option C here)

Note: I do think we could ship without any pagination at all, and add pagination in a separate PR

**Comment by gmajoulet** [comment]
If we keep the decision made here https://github.com/ampproject/amphtml/issues/29289 then I agree with Jon that option B is good.

I don't have strong opinions on attributes vs JSON configuration though, people will follow the documentation and both are easy to implement.

**Comment by Enriqe** [comment]
WDYT of joining this and the UI controls API (#30031) into one JSON configuration? 

I'm thinking of the following API:


```
<amp-story-player>
  <script type="application/json">
    {
      "behavior": {
        "on": "end", // for future-proofing
        "action": "fetch", // or circular-wrapping
        ...
      },
      "controls": {
        "header": { ... },
        ...
      }
    }
  </script>
</amp-story-player>
```

**Comment by Enriqe** [comment]
Closed by #30676
